src/MongoDB.Driver/MongoCollectionImpl.cs

```diff
@@ -806,19 +806,34 @@ private FindOperation<TResult> CreateAggregateToCollectionFindOperation<TResult>
                     break;
                 case "$merge":
                     {
-                        var mergeArguments = outStage[0].AsBsonDocument;
+                        var mergeArguments = outStage[0];
                         DatabaseNamespace outputDatabaseNamespace;
                         string outputCollectionName;
-                        var into = mergeArguments["into"];
-                        if (into.IsString)
+                        if (mergeArguments.IsString)
                         {
                             outputDatabaseNamespace = _collectionNamespace.DatabaseNamespace;
-                            outputCollectionName = into.AsString;
+                            outputCollectionName = mergeArguments.AsString;
                         }
                         else
                         {
-                            outputDatabaseNamespace = new DatabaseNamespace(into["db"].AsString);
-                            outputCollectionName = into["coll"].AsString;
+                            var into = mergeArguments.AsBsonDocument["into"];
+                            if (into.IsString)
+                            {
+                                outputDatabaseNamespace = _collectionNamespace.DatabaseNamespace;
+                                outputCollectionName = into.AsString;
+                            }
+                            else
+                            {
+                                if (into.AsBsonDocument.Contains("db"))
```

Comment:
consider:

    outputDatabaseNamespace  = into.AsBsonDocument.GetValue("db", _collectionNamespace.DatabaseNamespace)

Reply:
I am inclined to leave it as is, because otherwise the namespace retrieval would look like:
```
var outputDatabaseName = into.AsBsonDocument.GetValue("db", _collectionNamespace.DatabaseNamespace.DatabaseName);
outputDatabaseNamespace = new DatabaseNamespace(outputDatabaseName.AsString);
```
And extracting database name with the default path of assembling it again doesn't feel ideal either.